module/os/linux/zfs/zvol_os.c

```diff
@@ -794,8 +945,58 @@ zvol_alloc(dev_t dev, const char *name)
 	list_link_init(&zv->zv_next);
 	mutex_init(&zv->zv_state_lock, NULL, MUTEX_DEFAULT, NULL);
 
-#ifdef HAVE_SUBMIT_BIO_IN_BLOCK_DEVICE_OPERATIONS
-#ifdef HAVE_BLK_ALLOC_DISK
+	/*
+	 * Just to decode this #ifdef rat's nest a little...
+	 *
+	 * The block layer has 3 interfaces for getting BIOs:
+	 *
+	 * 1. blk-mq request queues (new)
+	 * 2. submit_bio() (oldest)
+	 * 3. regular request queues (old).
+	 *
+	 * Each of those interfaces has two permutations:
+	 *
+	 * a) We have blk_alloc_disk()/blk_mq_alloc_disk(), which allocates
+	 *    both the disk and its queue (5.14 kernel or newer)
+	 *
+	 * b) We don't have blk_*alloc_disk(), and have to allocate the
+	 *    disk and the queue separately. (5.13 kernel or older)
+	 */
+#if defined(HAVE_BLK_MQ)
+
+	/* Allocate our blk-mq tag_set */
+	if (zvol_blk_mq_alloc_tag_set(zv) != 0)
+		goto out_kmem;
+
+#if defined(HAVE_BLK_ALLOC_DISK)
+	zso->zvo_disk = blk_mq_alloc_disk(&zso->tag_set, zv);
+	if (zso->zvo_disk == NULL)
+		goto out_kmem;
+	zso->zvo_queue = zso->zvo_disk->queue;
+	zso->zvo_disk->minors = ZVOL_MINORS;
+#else
+	zso->zvo_disk = alloc_disk(ZVOL_MINORS);
+	if (zso->zvo_disk == NULL) {
+		blk_cleanup_queue(zso->zvo_queue);
+		goto out_kmem;
+	}
+	/* Allocate queue */
+	zso->zvo_queue = blk_mq_init_queue(&zso->tag_set);
+	if (IS_ERR(zso->zvo_queue)) {
+		blk_mq_free_tag_set(&zso->tag_set);
+		goto out_kmem;
+	}
+
+	/* Our queue is now created, assign it to our disk */
+	zso->zvo_disk->queue = zso->zvo_queue;
+
+#endif /* HAVE_BLK_ALLOC_DISK */
+
+	/* Finish blk-mq init */
+	blk_queue_logical_block_size(zso->zvo_queue, 4096);
```

Comment:
This needs to be `512` for compatibility with existing ZVOLs and any consumers expecting standard geometry.
Suggest using a dataset (zvol-specific) property for `logvolsize` and setting the `blk_queue_logical_block_size` off of that property.

Reply:
512 is the default logical_block_size. Looks like physical_block_size is expected to be equal or greater than logical so setting logical block size would impact the blk_queue_physical_block_size() called in zvol_os_create_minor() later.
```
/**
 * blk_queue_physical_block_size - set physical block size for the queue
 * @q:  the request queue for the device
 * @size:  the physical block size, in bytes
 *
 * Description:
 *   This should be set to the lowest possible sector size that the
 *   hardware can operate on without reverting to read-modify-write
 *   operations.
 */
void blk_queue_physical_block_size(struct request_queue *q, unsigned int size)
{
	q->limits.physical_block_size = size;

	if (q->limits.physical_block_size < q->limits.logical_block_size)
		q->limits.physical_block_size = q->limits.logical_block_size;

	if (q->limits.io_min < q->limits.physical_block_size)
		q->limits.io_min = q->limits.physical_block_size;
}
EXPORT_SYMBOL(blk_queue_physical_block_size);

```
I think we can remove/omit `blk_queue_logical_block_size()` and use default value.

Reply:
While i agree that allowing the default to work here should fix back-compat, it _is_ an opportunity to make ZVOLs more flexible in how they present to consumers by use of the `logvolblocksize` property (or even looking for a user-property for now if we dont want to bump the feature flags for that one change). Databases which work with 8K page sizes natively could benefit from this, as might storage virtualization which performs its own IO aggregation between the VM and backing store.